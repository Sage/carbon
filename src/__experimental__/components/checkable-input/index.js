export { default } from './checkable-input.component';
