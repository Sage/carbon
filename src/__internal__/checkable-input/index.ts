export { default } from "./checkable-input.component";
export type {
  CheckableInputProps,
  CommonCheckableInputProps,
} from "./checkable-input.component";
