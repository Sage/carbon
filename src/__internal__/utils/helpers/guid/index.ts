/**
 * Generates a random guid, useful for creating unique IDs.
 */
export default (): string => {
  function s4(): string {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  }
  return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
};
