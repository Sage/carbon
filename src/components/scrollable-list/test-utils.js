import React from 'react';
import { makeArrayKeys } from '../../__spec_helper__/test-utils';
import ScrollableListItem from './scrollable-list-item.component';

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArrayKeys(opts.num).reduce(listItemReducer(opts), []);
};

export default renderListItems;
