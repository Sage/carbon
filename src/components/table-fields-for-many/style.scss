@import "colors";

.ui-table-fields-for-many {
  border: 1px solid $border-color {
    collapse: separate;
    radius: 4px;
    spacing: 0;
  }
}
