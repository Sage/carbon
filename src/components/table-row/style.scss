@import "colors";

$row__background-color--even: $white;
$row__background-color--odd: #F2F2F2;
$row__background-color--hover: #D2EED1;
$row__border-radius: 4px;
$row__delete-icon-size: 18px;
$row__delete-icon-color--normal: #AFAFAF;
$row__delete-icon-color--hover: $error;

.ui-table-row {
  background-color: $row__background-color--odd;

  // first row
  &:first-child {
    .ui-table-row__td {
      // first td in first row
      &:first-child {
        border-radius: $row__border-radius 0 0 0;
      }

      // last td in first row
      &:last-child {
        border-radius: 0 $row__border-radius 0 0;
      }
    }
  }

  // last row
  &:last-child {
    .ui-table-row__td {
      // first td in last row
      &:first-child {
        border-radius: 0 0 0 $row__border-radius;
      }

      // last td in last row
      &:last-child {
        border-radius: 0 0 $row__border-radius 0;
      }
    }
  }

  // even rows
  &:nth-child(even) {
    background-color: $row__background-color--even;
  }

  // row hover
  &:hover {
    background-color: $row__background-color--hover;

    // delete icon on row hover
    .ui-table-row__delete-icon:before {
      content: "w";
      font-size: $row__delete-icon-size;
    }
  }
}

.ui-table-row__td {
  padding: 4px 2px;
}

.ui-table-row__delete {
  background: none;
  border: none;
  cursor: pointer;
  display: block;
  padding: 0;
  width: $row__delete-icon-size;
}

.ui-table-row__delete-icon {
  &:before {
    color: $row__delete-icon-color--normal;
    font-size: $row__delete-icon-size - 2;
  }

  &:hover:before {
    color: $row__delete-icon-color--hover;
  }
}
