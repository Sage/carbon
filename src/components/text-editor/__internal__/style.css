
.textBold {
  font-weight: bold;
}
.textItalic {
  font-style: italic;
}
