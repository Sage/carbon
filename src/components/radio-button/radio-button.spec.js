import React, { useRef } from "react";
import { mount } from "enzyme";
import { ThemeProvider } from "styled-components";
import TestRenderer from "react-test-renderer";
import { RadioButton, RadioButtonGroup } from ".";
import FieldHelpStyle from "../../__internal__/field-help/field-help.style";
import HiddenCheckableInputStyle from "../../__internal__/checkable-input/hidden-checkable-input.style";
import { StyledCheckableInput } from "../../__internal__/checkable-input/checkable-input.style";
import StyledCheckableInputSvgWrapper from "../../__internal__/checkable-input/checkable-input-svg-wrapper.style";
import { assertStyleMatch } from "../../__spec_helper__/test-utils";
import guid from "../../__internal__/utils/helpers/guid";
import baseTheme from "../../style/themes/base";
import mintTheme from "../../style/themes/mint";
import RadioButtonStyle from "./radio-button.style";
import Tooltip from "../tooltip";
import StyledHelp from "../help/help.style";

jest.mock("../../__internal__/utils/helpers/guid");
guid.mockImplementation(() => "guid-12345");

function render(props = {}, theme = mintTheme, renderer = mount) {
  const { error, info, warning, ...buttonProps } = props;
  const groupProps = {
    error,
    info,
    warning,
    name: "radio-button-name",
  };
  return renderer(
    <ThemeProvider theme={theme}>
      <RadioButtonGroup name="my-radio-group" {...groupProps}>
        <RadioButton name="my-radio" value="test" {...buttonProps} />
      </RadioButtonGroup>
    </ThemeProvider>
  );
}

const getRadioButton = (wrapper) => wrapper.find(RadioButton);

const validationTypes = ["error", "warning", "info"];

describe("RadioButton", () => {
  describe("propTypes", () => {
    it("does not allow a children prop", () => {
      spyOn(console, "error");
      render({ children: "someChildren" });
      const expected =
        "Forbidden prop `children` supplied to `RadioButton`. " +
        "This component is meant to be used as a self-closing tag. " +
        "You should probably use the label prop instead.";
      const actual = console.error.calls.argsFor(0)[0];
      expect(actual).toMatch(expected);
    });
  });

  it("the input ref should be forwarded", () => {
    let ref;
    const WrapperComponent = () => {
      ref = useRef();

      return <RadioButton name="my-radio" value="test" ref={ref} />;
    };
    const wrapper = mount(<WrapperComponent />);

    expect(ref.current).toEqual(
      wrapper.find(HiddenCheckableInputStyle).getDOMNode()
    );
  });

  describe("base", () => {
    it("renders as expected", () => {
      expect(
        render({}, mintTheme, TestRenderer.create).toJSON()
      ).toMatchSnapshot();
    });
  });

  describe("when disabled === true", () => {
    describe("default", () => {
      const wrapper = render({ disabled: true });

      it("disables the input", () => {
        const radioInput = wrapper.find("input");
        expect(radioInput.getDOMNode().disabled).toBe(true);
      });

      it("applies the correct circle styles", () => {
        assertStyleMatch(
          { fill: "var(--colorsCtilityDisabled400)" },
          getRadioButton(wrapper),
          { modifier: "circle" }
        );
      });

      it("renders the correct checked colour", () => {
        assertStyleMatch(
          { fill: "var(--colorsUtilityDisabled600)" },
          getRadioButton(wrapper),
          {
            modifier: `${`${HiddenCheckableInputStyle}:checked + ${StyledCheckableInputSvgWrapper} circle`}`,
          }
        );
      });
    });
  });

  describe('when size === "large"', () => {
    describe("default", () => {
      const wrapper = getRadioButton(render({ size: "large" }));
      const dimensions = { height: "24px", width: "24px" };

      it("applies the correct input styles", () => {
        assertStyleMatch({ ...dimensions }, wrapper, {
          modifier: `${StyledCheckableInput}`,
        });
      });

      it("applies the correct hidden input styles", () => {
        assertStyleMatch(dimensions, wrapper, {
          modifier: `${HiddenCheckableInputStyle}`,
        });
      });

      it("applies the correct svg wrapper styles", () => {
        assertStyleMatch(dimensions, wrapper, {
          modifier: `${StyledCheckableInputSvgWrapper}`,
        });
      });

      it("applies the correct svg styles", () => {
        assertStyleMatch(dimensions, wrapper, { modifier: "svg" });
      });

      it("applies the correct circle styles", () => {
        assertStyleMatch({ r: "3.75" }, wrapper, { modifier: "circle" });
      });
    });

    describe("and reverse === true", () => {
      describe("default", () => {
        const wrapper = getRadioButton(
          render({ reverse: true, size: "large" })
        );

        it("applies the correct FieldHelp styles", () => {
          assertStyleMatch({ padding: "0" }, wrapper, {
            modifier: `${FieldHelpStyle}`,
          });
        });
      });

      describe("and fieldHelpInline === true", () => {
        it("does not apply padding changes to FieldHelp", () => {
          const wrapper = render({
            fieldHelpInline: true,
            reverse: true,
            size: "large",
          });
          assertStyleMatch({ padding: undefined }, getRadioButton(wrapper), {
            modifier: `${FieldHelpStyle}`,
          });
        });
      });
    });
  });

  describe("validations", () => {
    describe.each(validationTypes)("%s === true", (type) => {
      it("show correct border on radio", () => {
        const wrapper = render({ [type]: true });
        const borderWidth = type === "error" ? 2 : 1;
        assertStyleMatch(
          {
            border: `${borderWidth}px solid ${baseTheme.colors[type]}`,
          },
          wrapper.find(RadioButton).at(0),
          { modifier: "svg" }
        );
      });
    });

    describe.each(validationTypes)('%s === "string"', (type) => {
      it("show correct border on radio", () => {
        const wrapper = render({ [type]: "Message" });
        const borderWidth = type === "error" ? 2 : 1;
        assertStyleMatch(
          {
            border: `${borderWidth}px solid ${baseTheme.colors[type]}`,
          },
          wrapper.find(RadioButton).at(0),
          { modifier: "svg" }
        );
      });
    });
  });

  describe("tooltipPosition", () => {
    it("overrides the default position when value is passed", () => {
      const { position } = mount(
        <RadioButton
          value="foo"
          label="foo"
          error="message"
          tooltipPosition="bottom"
        />
      )
        .find(Tooltip)
        .props();

      expect(position).toEqual("bottom");
    });
  });

  it("applies the correct Legend Container styles", () => {
    assertStyleMatch(
      {
        marginLeft: "32px",
      },
      mount(<RadioButtonStyle inline />),
      { modifier: "&:not(:first-of-type)" }
    );
  });

  describe("helpAriaLabel", () => {
    it("should set the aria-label on the Help component", () => {
      const text = "foo";

      const { "aria-label": ariaLabel } = mount(
        <RadioButton
          value="foo"
          label={text}
          labelHelp={text}
          helpAriaLabel={text}
        />
      )
        .find(StyledHelp)
        .props();

      expect(ariaLabel).toEqual(text);
    });
  });
});
