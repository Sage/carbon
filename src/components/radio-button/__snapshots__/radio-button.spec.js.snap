// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton base renders as expected 1`] = `
.c5 + .c4 {
  margin-top: 16px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c11 {
  cursor: pointer;
  opacity: 0;
  margin: 0;
  position: relative;
  z-index: 999;
}

.c9 {
  display: inline-block;
  position: relative;
}

.c13 .c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c13 .c10:hover,
.c13 .sc-ifAKCX:hover,
.c13 .c10:focus,
.c13 .sc-ifAKCX:focus {
  outline: none;
  cursor: not-allowed;
}

.c3 .c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 .c19 {
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding-top: 0;
  width: auto;
}

.c3 .c19 .c20,
.c3 .c19 .c21 {
  color: rgba(0,0,0,0.65);
  vertical-align: middle;
}

.c3 .c19 .c20:hover,
.c3 .c19 .c21:hover,
.c3 .c19 .c20:focus,
.c3 .c19 .c21:focus {
  color: rgba(0,0,0,0.9);
}

.c3 .c18 {
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c14 .c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c15 .c8 {
  padding-top: 1px;
}

.c15 .c12 {
  height: 16px;
}

.c15 .c10,
.c15 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c15 .c8,
.c15 .c10,
.c15 .c12,
.c15 svg {
  box-sizing: border-box;
  min-width: 16px;
  width: 16px;
}

.c15 .c10:not([disabled]):focus + .c12,
.c15 .c10:not([disabled]):hover + .c12 {
  box-shadow: 0 0 0 3px #FFB500;
}

.c15 .c10:checked ~ .c12 svg path {
  fill: rgba(0,0,0,0.90);
}

.c15 .c10:checked ~ .c12 svg path {
  fill: #CCD6DB;
}

.c15 .c12:hover,
.c15 .c12:focus {
  outline: none;
  cursor: not-allowed;
}

.c15.c15 .c4 {
  margin: 0;
}

.c15 .c12 {
  padding: 0;
}

.c15 .c12,
.c15 svg {
  border-radius: 50%;
}

.c15 .c8,
.c15 .c10,
.c15 .c12,
.c15 svg {
  height: 16px;
  width: 16px;
}

.c15 .c10:checked + .c12 circle {
  fill: var(--colorsUtilityYin090);
}

.c15 .c10:checked + .c12 circle {
  fill: var(--colorsUtilityDisabled600);
}

.c16 .c8 {
  padding-top: 1px;
}

.c16 .c12 {
  height: 16px;
}

.c16 .c10,
.c16 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c16 .c8,
.c16 .c10,
.c16 .c12,
.c16 svg {
  box-sizing: border-box;
  min-width: 16px;
  width: 16px;
}

.c16 .c10:not([disabled]):focus + .c12,
.c16 .c10:not([disabled]):hover + .c12 {
  box-shadow: 0 0 0 3px #FFB500;
}

.c16 .c12 {
  height: 24px;
}

.c16 .c8,
.c16 .c10,
.c16 .c12,
.c16 svg {
  height: 24px;
  width: 24px;
}

.c16 .c10:checked ~ .c12 svg path {
  fill: rgba(0,0,0,0.90);
}

.c16.c16 .c4 {
  margin: 0;
}

.c16 .c12 {
  padding: 0;
}

.c16 .c12,
.c16 svg {
  border-radius: 50%;
}

.c16 .c8,
.c16 .c10,
.c16 .c12,
.c16 svg {
  height: 16px;
  width: 16px;
}

.c16 .c10:checked + .c12 circle {
  fill: var(--colorsUtilityYin090);
}

.c16 .c8,
.c16 .c10,
.c16 .c12,
.c16 svg {
  height: 24px;
  width: 24px;
}

.c17 .c8 {
  padding-top: 1px;
}

.c17 .c12 {
  height: 16px;
}

.c17 .c10,
.c17 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c17 .c8,
.c17 .c10,
.c17 .c12,
.c17 svg {
  box-sizing: border-box;
  min-width: 16px;
  width: 16px;
}

.c17 .c10:not([disabled]):focus + .c12,
.c17 .c10:not([disabled]):hover + .c12 {
  box-shadow: 0 0 0 3px #FFB500;
}

.c17 .c12 {
  height: 24px;
}

.c17 .c8,
.c17 .c10,
.c17 .c12,
.c17 svg {
  height: 24px;
  width: 24px;
}

.c17 .c10:checked ~ .c12 svg path {
  fill: rgba(0,0,0,0.90);
}

.c17.c17 .c4 {
  margin: 0;
}

.c17 .c12 {
  padding: 0;
}

.c17 .c12,
.c17 svg {
  border-radius: 50%;
}

.c17 .c8,
.c17 .c10,
.c17 .c12,
.c17 svg {
  height: 16px;
  width: 16px;
}

.c17 .c10:checked + .c12 circle {
  fill: var(--colorsUtilityYin090);
}

.c17 .c8,
.c17 .c10,
.c17 .c12,
.c17 svg {
  height: 24px;
  width: 24px;
}

.c2 {
  margin-bottom: var(--spacing150);
}

.c2 .c8 {
  padding-top: 1px;
}

.c2 .c12 {
  height: 16px;
}

.c2 svg {
  background-color: #FFFFFF;
  border: 1px solid #668592;
}

.c2 .c10,
.c2 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c2 .c8,
.c2 .c10,
.c2 .c12,
.c2 svg {
  box-sizing: border-box;
  min-width: 16px;
  width: 16px;
}

.c2 .c10:not([disabled]):focus + .c12,
.c2 .c10:not([disabled]):hover + .c12 {
  box-shadow: 0 0 0 3px #FFB500;
}

.c2 .c19 {
  width: auto;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
}

.c2 .c18 {
  margin-left: 16px;
  margin-top: 0;
  padding-left: 8px;
}

.c2 .c21 {
  position: relative;
  display: inline-block;
}

.c2 .c10:checked ~ .c12 svg path {
  fill: rgba(0,0,0,0.90);
}

.c2:last-of-type {
  margin-bottom: 0;
}

.c2.c2 .c4 {
  margin: 0;
}

.c2 .c12 {
  padding: 0;
}

.c2 .c12,
.c2 svg {
  border-radius: 50%;
}

.c2 .c8,
.c2 .c10,
.c2 .c12,
.c2 svg {
  height: 16px;
  width: 16px;
}

.c2 svg {
  padding: 1px;
}

.c2 circle {
  r: 5;
}

.c2 .c19 {
  -webkit-flex: 1 1 calc(100% - 44px);
  -ms-flex: 1 1 calc(100% - 44px);
  flex: 1 1 calc(100% - 44px);
}

.c2 .c10:checked + .c12 circle {
  fill: var(--colorsUtilityYin090);
}

.c0 {
  margin: 0;
  border: none;
  padding: 0;
  min-width: 0;
  min-inline-size: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<fieldset
  className="c0"
  data-component="radiogroup"
>
  <div
    className="c1"
    data-component="radio-button-group"
    role="radiogroup"
  >
    <div
      className="c2"
      data-component="radio-button"
    >
      <div
        className="c3"
      >
        <div
          className="c4 c5"
        >
          <div
            className="c6 c7"
          >
            <div
              className="c8 c9"
            >
              <input
                aria-checked={false}
                aria-describedby=""
                aria-invalid={false}
                checked={false}
                className="c10 c11"
                id="guid-12345"
                name="radio-button-name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                role="radio"
                type="radio"
                value="test"
              />
              <div
                className="c12 "
              >
                <svg
                  focusable="false"
                  viewBox="0 0 15 15"
                >
                  <g
                    fill="none"
                    fillRule="evenodd"
                    stroke="none"
                    strokeWidth="1"
                  >
                    <circle
                      className="radio-button-check"
                      cx="7.5"
                      cy="7.5"
                      fill="#FFFFFF"
                      r="5"
                    />
                  </g>
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;
